testsuite/cluster-test/src/atomic_histogram.rs

```diff
@@ -0,0 +1,188 @@
+}
+
+impl Default for AtomicHistogramAccumulator {
+    fn default() -> AtomicHistogramAccumulator {
+        AtomicHistogramAccumulator::new(DEFAULT_HISTOGRAM_CAPACITY, DEFAULT_HISTOGRAM_STEP_WIDTH)
+    }
+}
+
+impl AtomicHistogramAccumulator {
+    pub fn new(size: usize, step: u64) -> AtomicHistogramAccumulator {
+        let mut buf = Vec::with_capacity(size);
+        for _i in 0..size {
+            buf.push(AtomicU64::new(0));
+        }
+        Self {
+            capacity: size,
+            step_width: step,
+            buckets: buf,
+        }
+    }
+
+    pub fn snapshot(&self) -> AtomicHistogramSnapshot {
+        let mut buf = Vec::with_capacity(self.capacity);
+        for i in 0..self.capacity {
+            buf.push(self.buckets[i].load(Ordering::Relaxed));
+        }
+        AtomicHistogramSnapshot {
+            capacity: self.capacity,
+            step_width: self.step_width,
+            buckets: buf,
+        }
+    }
+
+    fn get_bucket_num(&self, data_value: u64) -> u64 {
+        let bucket_num = data_value / self.step_width;
+        if bucket_num >= self.capacity as u64 - 2 {
+            return self.capacity as u64 - 1;
+        }
+        bucket_num
+    }
+
+    pub fn record_data_point(&self, data_value: u64, data_num: u64) {
+        let bucket_num = self.get_bucket_num(data_value);
+        self.buckets[bucket_num as usize].fetch_add(data_num as u64, Ordering::Relaxed);
+    }
+}
+
+#[derive(Debug)]
+pub struct AtomicHistogramSnapshot {
+    capacity: usize,
+    step_width: u64,
+    buckets: Vec<u64>,
+}
+
+impl Default for AtomicHistogramSnapshot {
+    fn default() -> AtomicHistogramSnapshot {
+        AtomicHistogramAccumulator::default().snapshot()
+    }
+}
+
+impl Sub for &AtomicHistogramSnapshot {
+    type Output = AtomicHistogramSnapshot;
+
+    fn sub(self, other: &AtomicHistogramSnapshot) -> AtomicHistogramSnapshot {
+        assert_eq!(
+            self.buckets.len(),
+            other.buckets.len(),
+            "Histogram snapshots must have same size, prev: {}, cur: {}",
+            self.buckets.len(),
+            other.buckets.len()
+        );
+        let mut buf = Vec::with_capacity(self.capacity);
+        for i in 0..self.buckets.len() {
+            buf.push(self.buckets[i] - other.buckets[i]);
+        }
+        AtomicHistogramSnapshot {
+            capacity: self.capacity,
+            step_width: self.step_width,
+            buckets: buf,
+        }
+    }
+}
+
+impl AtomicHistogramSnapshot {
+    pub fn cal_percentile_latency(
+        &self,
+        numerator: u64,
+        denominator: u64,
+    ) -> u64 {
+        let committed: u64 = self.buckets.iter().sum();
+        let p_count = committed * numerator / denominator;
+        let mut counter = 0u64;
+        for i in 0..self.buckets.len() {
+            counter += self.buckets[i];
+            if counter >= p_count {
+                return i as u64 * self.step_width;
+            }
+        }
+        0
```

Comment:
Should you use `unreachable!()` here? It should not be possible to enter this condition(and if it is possible, 0 is certainly not a valid return here)